src/GovUk.Education.ExploreEducationStatistics.Data.Services/SubjectMetaService.cs

```diff
@@ -335,5 +333,114 @@ private static string OrderLocationAttributes(LocationAttributeNode node)
+                .OnSuccess(_ =>
+                {
+                    var requestMap = requestFilterGroups.ToDictionary(filterGroup => filterGroup.Id);
+                    var results = filter.FilterGroups.Select(filterGroup =>
+                            AssertCollectionsAreSameIgnoringOrder(filterGroup.FilterItems,
+                                requestMap[filterGroup.Id].FilterItems,
+                                indicator => indicator.Id,
+                                $"Requested filter items do not match subject for filter group: {filterGroup.Id}"))
+                        .ToList();
+
+                    var failure = results.FirstOrDefault(result => result.IsLeft)?.Left;
+                    return failure ?? new Either<ActionResult, Unit>(Unit.Instance);
+                });
+        }
+
+        private async Task<Either<ActionResult, Unit>> ValidateIndicatorGroupsForSubject(
+            Guid subjectId,
+            List<IndicatorGroupUpdateViewModel> requestIndicatorGroups)
+        {
+            var indicatorGroups = await _indicatorGroupRepository.GetIndicatorGroups(subjectId);
+            return AssertCollectionsAreSameIgnoringOrder(indicatorGroups,
+                    requestIndicatorGroups,
+                    indicatorGroup => indicatorGroup.Id,
+                    requestIndicatorGroup => requestIndicatorGroup.Id,
+                    "Requested indicator groups do not match subject indicator groups")
+                .OnSuccess(_ =>
+                {
+                    var requestMap = requestIndicatorGroups.ToDictionary(indicatorGroup => indicatorGroup.Id);
+                    var results = indicatorGroups.Select(indicatorGroup =>
+                            AssertCollectionsAreSameIgnoringOrder(indicatorGroup.Indicators,
+                                requestMap[indicatorGroup.Id].Indicators,
+                                indicator => indicator.Id,
+                                $"Requested indicators do not match subject for indicator group: {indicatorGroup.Id}"))
+                        .ToList();
+
+                    var failure = results.FirstOrDefault(result => result.IsLeft)?.Left;
+                    return failure ?? new Either<ActionResult, Unit>(Unit.Instance);
+                });
+        }
+
+        private static Either<ActionResult, Unit> AssertCollectionsAreSameIgnoringOrder<TLeft, TRight, TId>(
+            IEnumerable<TLeft> left,
+            IEnumerable<TRight> right,
+            Func<TLeft, TId> leftIdSelector,
+            Func<TRight, TId> rightIdSelector,
+            string error)
+        {
+            var rightIdList = right.Select(rightIdSelector).ToList();
+            return AssertCollectionsAreSameIgnoringOrder(left, rightIdList, leftIdSelector, error);
+        }
+
+        private static Either<ActionResult, Unit> AssertCollectionsAreSameIgnoringOrder<TLeft, TId>(
+            IEnumerable<TLeft> left,
+            IEnumerable<TId> right,
```

Comment:
Why do we have one of these lists as an Id list and the other as entities with an Id selector?

Reply:
It was for convenience because in a couple of cases we want to compare a list of either filter items or indicators, with a list of Id's that are supplied in the request.

In the other cases we want to compare two lists of entities where the lists are not the same type and supply Id selectors for both.

I've simplified things a bit now so we don't have this overload anymore. Now we either make the comparison with two Id lists or two entity lists with Id selectors.

Reply:
Nice one, I see.  Cool yeah I think a bit simpler is probably better in the long run.